### ISSUE TITLE
Improve the performance of ListingTable

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#9964]
### Is your feature request related to a problem or challenge?

The `ListingTable`  works quite well in practice, but like all software could be made better. I am writing up this ticket to enumerate some areas for improvement in the hopes people who are interested can collaborate / coordinate their efforts

### Background

DataFusion has a [`ListingTable`](https://github.com/apache/arrow-datafusion/blob/2dad90425bacb98a3c2a4214faad53850c93104e/datafusion/core/src/datasource/listing/table.rs#L443-L508) that effectively reading tables stored in one or more files in a "hive partitioned" directory structure:

So for example, give files like this:
```
/path/to/my_table/file1.parquet
/path/to/my_table/file2.parquet
/path/to/my_table/file3.parquet
```

You can create a table with a command like

```sql
CREATE EXTERNAL TABLE my_table
LOCATION '/path/to/my_table'
```

And the `ListingTable` will handle figuring out schema, and running queries against those files as though they were a single table. 

### Describe the solution you'd like

Here are some things I suspect could be improved:

### Object store list caching
For large tables (many files) on remote stores, the actual object store call to [`LIST`](https://github.com/apache/arrow-datafusion/blob/2b0a7db0ce64950864e07edaddfa80756fe0ffd5/datafusion/core/src/datasource/listing/url.rs#L218) may be non trivially expensive and thus perhaps could be cached

### MetaData caching 

`ListingTable` ([code link](https://github.com/apache/arrow-datafusion/blob/2dad90425bacb98a3c2a4214faad53850c93104e/datafusion/core/src/datasource/listing/table.rs#L829-L857))  prunes files based on statistics, and then inside the ParquetExec itself ([link](https://github.com/apache/arrow-datafusion/blob/2dad90425bacb98a3c2a4214faad53850c93104e/datafusion/core/src/datasource/physical_plan/parquet/mod.rs#L527-L569)) where it again prunes row groups and data pages based on metadata,

###  IO granularity 

I have heard it said that the DataFusion `ParquetExec` reader reads a page at a time -- this is fine if the parquet file is a local file on disk, but it is likely quite inefficient if each page must be fetched with an individual remote object store request. This assertion needs to be researched, but if true we could make queries on remote parquet files much faster by making fewer larger requests

### Describe alternatives you've considered

@Ted-Jiang  added some APIs in https://github.com/apache/arrow-datafusion/pull/7570 https://github.com/apache/arrow-datafusion/blob/2b0a7db0ce64950864e07edaddfa80756fe0ffd5/datafusion/execution/src/cache/mod.rs but there aren't any default implementations in DataFusion so the metadata is read multiple times

Maybe we can add a default implementation of the caches in SessionContext with a simple policy (like LRU / some max size)

Another potential way to improve performance is to cache the decoded metadata from the Parquet footer rather than checking it once to prune files and then again to prune row groups / pages

### Additional context

@matthewmturner mentioned interest in improving listing table performance: https://github.com/apache/arrow-datafusion/issues/9899#issuecomment-2030139830

Note we don't use ListingTable in InfluxDB for some of the reasons described above

**Comment by Lordworms** [comment]
Interested in this one

**Comment by alamb** [comment]
FYI I think this is more like an Epic that can be used to coordinate individual tasks / changes rather than a specific change itself.

>  Interested in this one

Thanks @Lordworms  -- one thing that would probably help to start this project along would be to gather some data.

Specifically, put the LIstingTable against data on a remote object store (eg. figure out how to write a query against 100 parquet files on an S3 bucket).

And then measure how much time is spent:
1. object store listing
2. fetching metadata
3. pruning / fetching IO
4. How many object store requests are made

Does anyone know a good public data set on S3 that we could use to test / benchmark with?

**Comment by Lordworms** [comment]
> FYI I think this is more like an Epic that can be used to coordinate individual tasks / changes rather than a specific change itself.
> 
> > Interested in this one
> 
> Thanks @Lordworms -- one thing that would probably help to start this project along would be to gather some data.
> 
> Specifically, put the LIstingTable against data on a remote object store (eg. figure out how to write a query against 100 parquet files on an S3 bucket).
> 
> And then measure how much time is spent:
> 
> 1. object store listing
> 2. fetching metadata
> 3. pruning / fetching IO
> 4. How many object store requests are made
> 
> Does anyone know a good public data set on S3 that we could use to test / benchmark with?

I got it, I'll search some data first

**Comment by Lordworms** [comment]
I'll start with a random dataset here https://registry.opendata.aws/

**Comment by MohamedAbdeen21** [comment]
(It's not a "performance" issue, but rather for better user experience. Also, it requires upstream changes to arrow-rs/object_store.)

The recent change in #9912  uses 10 random files to infer the partition columns, this means that we may fail to catch corrupted/manually-changed partitions on table creation (shouldn't be a common case). This is because `ObjectStore` only provides `list` function to retrieve objects.

If we can provide a BFS approach to traverse the Object Store and use that in partition inference, that will be a nice QoL change. 

Curious to know how we feel about upstream changes for such non-critical changes in DF?

**Comment by alamb** [comment]
> The recent change in https://github.com/apache/arrow-datafusion/pull/9912 uses 10 random files to infer the partition columns, this means that we may fail to catch corrupted/manually-changed partitions on table creation (shouldn't be a common case). This is because ObjectStore only provides list function to retrieve objects.

I think `ObjectStore` only provides a LIST api as that is the functionality offered by S3, GCP, etc. 

> The recent change in https://github.com/apache/arrow-datafusion/pull/9912 uses 10 random files to infer the partition columns, this means that we may fail to catch corrupted/manually-changed partitions on table creation (shouldn't be a common case).

In this particular case, I think there are different needs for different users (some might want to pay the cost for a full validation, but some might be happy with just a sanity check) One approach would be to add a config option to DataFusion that controls the maxumum number of paths to check when creating a listing table 🤔 

> Curious to know how we feel about upstream changes for such non-critical changes in DF?

In general I think upstream changes are great to propose if they help more than just DataFusion

**Comment by Lordworms** [comment]
I have done some basic play with the bitcoin dataset
![17347f2f94015d8396ec20a0817a6f09](https://github.com/apache/arrow-datafusion/assets/48054792/84beed34-f1f9-4f3f-b485-c7a312a9778f)
and also did some profiling with instrument


> FYI I think this is more like an Epic that can be used to coordinate individual tasks / changes rather than a specific change itself.
> 
> > Interested in this one
> 
> Thanks @Lordworms -- one thing that would probably help to start this project along would be to gather some data.
> 
> Specifically, put the LIstingTable against data on a remote object store (eg. figure out how to write a query against 100 parquet files on an S3 bucket).
> 
> And then measure how much time is spent:
> 
> 1. object store listing
<img width="1202" alt="image" src="https://github.com/apache/arrow-datafusion/assets/48054792/0efbf812-3ed5-4603-8ee5-9fbe5a1b365b">
> 2. fetching metadata
<img width="1175" alt="image" src="https://github.com/apache/arrow-datafusion/assets/48054792/2fca61e9-05c7-4366-8b6f-a72c8d80f6dc">

> 3. pruning / fetching IO
> 4. How many object store requests are made
> 
> Does anyone know a good public data set on S3 that we could use to test / benchmark with?

just want to know what is a good start to solving this issue, should I implement the cache https://github.com/apache/arrow-datafusion/blob/2b0a7db0ce64950864e07edaddfa80756fe0ffd5/datafusion/execution/src/cache/mod.rs here first?

**Comment by alamb** [comment]
> And then measure how much time is spent:

that is very interesting

> just want to know what is a good start to solving this issue, should I implement the cache

just want to know what is a good start to solving this issue, should I implement the cache https://github.com/apache/arrow-datafusion/blob/2b0a7db0ce64950864e07edaddfa80756fe0ffd5/datafusion/execution/src/cache/mod.rs here first?

If indeed most of the exection time is spent parsing (or fetching) parquet metadata, implementing a basic cache would likely help.

Also, @tustvold  brought https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/parquet/trait.ParquetFileReaderFactory.html to my attention which might be able to help avoid the overhead

So what I suggest is:
1. Do a proof of concept (POC - hack it in, don't worry about tests, etc) with your approach and see if you can show performance improvements (https://github.com/apache/arrow-datafusion/pull/9708  is an example of such a PR)
2. If you can show it improves performance significantly, then we can work on a final design / tests / etc

The reason to do the POC first is that performance analysis is notoriously tricky at the system lavel so you want to have evidence your work will actually improve performance before you spend a bunch of time polishing up the PR  (it is very demotivating, at least to me, to make a beautiful PR only to find out it doesn't really help performance)

**Comment by Lordworms** [comment]
> > And then measure how much time is spent:
> 
> that is very interesting
> 
> > just want to know what is a good start to solving this issue, should I implement the cache
> 
> just want to know what is a good start to solving this issue, should I implement the cache https://github.com/apache/arrow-datafusion/blob/2b0a7db0ce64950864e07edaddfa80756fe0ffd5/datafusion/execution/src/cache/mod.rs here first?
> 
> If indeed most of the exection time is spent parsing (or fetching) parquet metadata, implementing a basic cache would likely help.
> 
> Also, @tustvold brought https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/parquet/trait.ParquetFileReaderFactory.html to my attention which might be able to help avoid the overhead
> 
> So what I suggest is:
> 
> 1. Do a proof of concept (POC - hack it in, don't worry about tests, etc) with your approach and see if you can show performance improvements ([WIP: Avoid copying LogicalPlans / Exprs during OptimizerPasses #9708](https://github.com/apache/arrow-datafusion/pull/9708)  is an example of such a PR)
> 2. If you can show it improves performance significantly, then we can work on a final design / tests / etc
> 
> The reason to do the POC first is that performance analysis is notoriously tricky at the system lavel so you want to have evidence your work will actually improve performance before you spend a bunch of time polishing up the PR (it is very demotivating, at least to me, to make a beautiful PR only to find out it doesn't really help performance)

Got it

**Comment by Lordworms** [comment]
Sorry for the late update, for a deeper analysis, I think a direct call to get ObjectStore (direct call, no pruning) could not be optimized by caching the MetaData. Currently, we mainly use two APIs from S3, the [list_object](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html) and the [get_object](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html) I looked into the code and did analysis with WireShark and find out current data fusion just do one list_object call and x(number of file) get_object call. I don't think these could be avoided by adding caches since in [fetch_parquet_metadata](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/src/datasource/file_format/parquet.rs#L389-L401) we need the actual parquet file body instead of ObjectMeta. I'll check the pruning issue to see if I can make some effort.

**Comment by matthewmturner** [comment]
@Lordworms if i recall correctly, the s3 list call is made on every query, and if the number of files is large this can be non-trivial, so if the listed files can be cached after creating / the first query then we could save that listing time on subsequent queries.

**Comment by Lordworms** [comment]
> @Lordworms if i recall correctly, the s3 list call is made on every query, and if the number of files is large this can be non-trivial, so if the listed files can be cached after creating / the first query then we could save that listing time on subsequent queries.

Yeah, I got it, I am implementing it, but I am confused about the caching granularity. whether to cache the whole parquet file or a portion of the file.

**Comment by matthewmturner** [comment]
I think both could be done (we actually do that) - but starting with the metadata to me would be a good place to start as to my understanding the builtin cache would be better suited for something like object metadata than the whole parquet files.

**Comment by Lordworms** [comment]
> but starting with the metadata to me would be a good place to start as to my understanding the builtin cache would be better suited for something like object metadata than the whole parquet files.

Sure, I got it. Thanks for the advice

**Comment by matthewmturner** [comment]
One additional comment, I'm not sure how the existing cache would perform for range queries. For example we a use trie in our implementation. 

Depending on your findings that could be a follow on though as there might still be enough benefit from a baseline implementation.

**Comment by Lordworms** [comment]
> One additional comment, I'm not sure how the existing cache would perform for range queries. For example we a use trie in our implementation.
> 
> Depending on your findings that could be a follow-up on though as there might still be enough benefit from a baseline implementation.

I was testing directly using 
<img width="704" alt="image" src="https://github.com/apache/arrow-datafusion/assets/48054792/a08b9068-fea6-46d3-bd82-3e429b3cc260">
and the performance has risen up a reasonable degree. I'll add  LRU logic and file granularity tomorrow(since the current logic just support directory level)

**Comment by Lordworms** [comment]
> One additional comment, I'm not sure how the existing cache would perform for range queries. For example we a use trie in our implementation.
> 
> Depending on your findings that could be a follow on though as there might still be enough benefit from a baseline implementation.

Yes, I think the basic implementation could not handle the both directory and file level. I may implement a trie or other BST structure

**Comment by Lordworms** [comment]
I have implemented a basic LRU metadata cache, and I think just caching the metadata would get slight performance improvement(we call the List_Object API just once but call the Get_Object API more than x times which x is the number of files), most of the time consumed should also be calling [get_object](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html) API since we should call it for each object. So I think caching the parquet file should be a better way since most of the time consumed is calling the get_range function(which calls the GetObject API and then reads the range of the parquet file)
1. Should we do the caching for the parquet file?  I've seen the current logic of 
2. How to implement the file cache, the current logic is complicated since not just the datafusion but the arrow-rs is calling get_range for example, in arrow readMetaData, it would call the  get_range API
<img width="887" alt="image" src="https://github.com/apache/arrow-datafusion/assets/48054792/972476da-e307-42a0-8d39-7143212c67da"> also should we cache the whole parquet file or parts of it?
3. which data structure should be chosen? I have tried the Sequence Trie and the LRU-DashMap.

I have been confused for the last two days and really appreciate your help @alamb @matthewmturner

**Comment by matthewmturner** [comment]
@Lordworms thanks for the work on this.

Just to confirm - what was the improvement in milliseconds we saw from the object meta cache?  For context, in low latency use cases (we have that requirement) even shaving off a few ms can be important so would still want to get that improvement in.

IMO it would be good to close the object meta cache topic first before moving on to file cache - unless there is some connection there that I am missing where we need to consider them together.

**Comment by Lordworms** [comment]
> @Lordworms thanks for the work on this.
> 
> Just to confirm - what was the improvement in milliseconds we saw from the object meta cache? For context, in low latency use cases (we have that requirement) even shaving off a few ms can be important so would still want to get that improvement in.
> 
> IMO it would be good to close the object meta cache topic first before moving on to file cache - unless there is some connection there that I am missing where we need to consider them together.

Sure, from left to right: unchanged, SequenceTrie cached, ThreadSafeLRU DashMap cached
![c55455d0e3b0fd70096706ad692cc162](https://github.com/apache/arrow-datafusion/assets/48054792/e4134a61-10bb-4c46-900a-f6e76bbeceda)

**Comment by matthewmturner** [comment]
Thats interesting that dashmap performed better.  Would you mind also doing a query with a filter / pruning involved and comparing the results (perhaps with a range scan as well)? And just to confirm, the cache is filled at creation time during schema inference, right?

Also is the cache used by both `ListingTable` and `ParquetExec`?  Im not sure how it works if adding the cache removes the duplicate calls or if additional work is required for that (if it's additional work could probably be done as a separate PR once cache implementation is finalized).

I think it would also be nice to have a new benchmark for this to help monitor moving forward.

**Comment by matthewmturner** [comment]
@Lordworms apologies it took me longer than expected to get some free time.  I plan to review between tonight and tomorrow.

**Comment by matthewmturner** [comment]
@Lordworms did you get the chance to compare querying with a filter / pruning involved (ideally with a range) between dashmap and sequence trie? Not sure if the dataset is conducive to that though.

**Comment by Lordworms** [comment]
> @Lordworms did you get the chance to compare querying with a filter / pruning involved (ideally with a range) between dashmap and sequence trie? Not sure if the dataset is conducive to that though.

I haven't added filter test cases yet, but I can do that in the following two days, I'll give you an update then.

**Comment by Lordworms** [comment]
Sorry for the late update, I was traveling last week, I'll test it today and give you feedback

**Comment by alamb** [comment]
We have made some progress on this recently so I am going to close down this old epic. 

I have filed a new planning epic for future improvements
- https://github.com/apache/datafusion/issues/17214
